Show diff in error msg when repr is long.

CharString opened the pull request:
Cleaned up #18 into a single patch.

Commit:
Show diff in error msg when repr is long.

Commit:
Fixes exception comparing bytes in py3

Result (failing/passing) should stay the same across pythons, but the
message on the AssertionError differs from 2 to 3.

So result is always tested, failmsg only for py3.

CharString commented:
Added a bugfix; expecting bytes in py3 would raise exceptions.

Commit:
Fix byte test for py2.6

sys.version_info only became a namedtuple in 2.7, using positional
reference instead.

garybernhardt commented:
Seems like a good change. My only worry is the py3k bytes test. It's big with lots of complex data in it, and the short circuiting conditional is vexing me. :)

With this change, expecter only tries to diff either known string types or the output of pformat. I don't think that the bytes-specific test is necessary as long as we can assume:
1. pformat will return a string
2. difflib won't blow up when given strings

Those seem like safe assumptions to me. If so, I think that the bytes test can just be deleted. But if I'm missing something (very possible), fill me in!

sigmavirus24 commented on expecter.py:
```diff
@@ -7,7 +8,7 @@
 
 
-basestring = getattr(__builtins__, 'basestring', (str, bytes))
+basestring = getattr(__builtins__, 'basestring', str)
```
I'm confused, why are you removing bytes?

CharString replied:
In python 2 str usage is mixed... Sometimes it's contains a naive string in the same encoding as the source file, sometimes it contains bytes. And then there's the unicode type.

In python 3 str is really just a unicode string type that contain codepoints and bytes really are just bytes. So to treat them as strings in difflib we must either guess the encoding or face `TypeError: string argument without an encoding` as  a2671f07c6055420c02e5c1e61db02e761f6d858 shows.

Now pformat doesn't guess the encoding bytes, but shows the bytevalues that could be interpreted in ASCII as chars, and the rest as two nibbles in hex escaped by `\x` (aka calling repr on a byte)... Calling repr on a byte makes much more sense.

Perhaps a even more sense would be showing just 12 pairs of nibbles separated by spaces on a line, like hexeditors do. And passing _that_ to difflib.

CharString commented:
@garybernhardt, the shortcircuit  on the message reeks and could be removed... Or we could change message for bytes in py3 into something like

```
Expected other bytes
Diff:
@@ -1 +1 @@
- 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f 00
+ 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f
```

But then again. When you're dealing with bytes, chances are you are using something like structlib in your expectations anyway, to keep you sane.

CharString commented:
Unless it's the 1st of April _and_ actual contains b'[\xf0\x9f\x92\xa9](http://codepoints.net/U+1F4A9)', then I feel inclined that the bytes are in fact utf-8 encoded.

sigmavirus24 commented:
> When you're dealing with bytes, chances are you are using something like structlib in your expectations anyway, to keep you sane.

That's a pretty heavy assumption and one which frankly doesn't float with me.

garybernhardt commented:
For this PR, I think it's fine to just diff the bytes' repr and remove the bytes-specific test. If bytes repeatedly become an issue in practice, that can be a separate PR. I'd rather avoid implementing such a thing preemptively. :)

CharString commented:
@garybernhardt should I reduce the test to

```
def it_can_compare_bytes(self):
    null = bytes((0,))
    expect(null) == null
    data = bytes(range(9, 32))
    def _fails():
        expect(data) == data + null
    assert_raises(AssertionError, _fails)
```

to prevent a regression? Or should it be removed all together?

garybernhardt commented:
I'm tempted to say that even that test isn't needed, since `it_shows_diff_for_large_reprs` already tests repr diffing, and we can trust repr and the stdlib. However, the basestring/str/bytes stuff is so error-prone, and there are so many subtle potential failure paths, that leaving a test in is probably a good idea. Push that simplified test and I'll merge this guy.

Unrelated to that, thanks for making that `split('\n')` -> `splitlines()` change. I'm embarrassed at having missed that!

Commit:
Cleanup comparing bytes test

Removing failmsg assertions, stdlib and repr can be trusted.
Leaving assertions that comparing bytes does function.

CharString commented:
Done. It'd be nice if the version on pypi got updated. I noticed the current one there doesn't even pass arguments to custom expectations.

On the `split('\n')`, I was to blame for that one in the first place. :$ 2e3c9058dc31d1a41380d25a8200ef0768dd183b

garybernhardt commented:
Merged. Thanks!

CharString commented:
Thanks for sharing this code.

garybernhardt commented:
Expecter 0.3.0 is pushed to PyPI!

CharString commented:
On 25 June 2014 20:36:47 CEST, Gary Bernhardt notifications@github.com wrote:

> Expecter 0.3.0 is pushed to PyPI!

W00T!!

Follow Gary's lead, everybody: get your partyhats on!!

(for the e-archaeologists of y3k: it's a reference to his profile picture... now that I got your attention, is python 2.7 still in use???)
